### ISSUE TITLE
Random reaction time for Application Failover 

### ISSUE THREAD

**Issue by lorenzogio97** [karmada-io/karmada#4208]
Hi,  
**What happened**:
i was trying to test the Application failover feature of Karmada, using a simple nginx deployment and a ClusterPropagationPolicy. I'm testing the exact same scenario reported here (https://karmada.io/docs/userguide/failover/application-failover/), but when I cordon the cluster where pods are currenty scheduled and delete them, karmada continue to try to re-schedule them in the same cluster, for more than the  tolerationSeconds that are expected. 
I made some tests with different tolerationSeconds value and I experience the following time to see the pods scheduled on the other cluster:
- 5s: 2 independent tests, 19 min and 16 min
- 8s: 12 min
- 15s: 29 min
From what I experienced, the time is not deterministic and i didn't see any relationship with tolerantSeconds value. 

**What you expected to happen**:
Pods scheduled (and then deleted, like it's done in doc) on the cordoned cluster should be re-scheduled on the other cluster after tolerationSeconds time value. 

**How to reproduce it (as minimally and precisely as possible)**:
- Deploy on karmada control plane this Deployment and ClusterPropagationPolicy (of course, change cluster names): 
[deployment.txt](https://github.com/karmada-io/karmada/files/13257280/deployment.txt)
[propagationpolicy.txt](https://github.com/karmada-io/karmada/files/13257281/propagationpolicy.txt)
- Cordon the primary cluster
- Delete pod on that cluster
- Wait that pods are re-scheduled on the backup cluster

**Anything else we need to know?**:
-

**Environment**:
- Karmada version: Karmada v1.7.0
- kubectl-karmada or karmadactl version (the result of `kubectl-karmada version` or `karmadactl version`): - 
- Others: The application has been deployed on Karmada control plane using ArgoCD

**Comment by jwcesign** [comment]
Hi @lorenzogio97 
I tried with following yaml with `hack/local-up-karmada.sh`:
```yaml
apiVersion: policy.karmada.io/v1alpha1
kind: ClusterPropagationPolicy
metadata:
  name: nginx-propagation
spec:
  propagateDeps: true
  failover:
    application:
      decisionConditions:
        tolerationSeconds: 5
      purgeMode: Immediately
  resourceSelectors:
    - apiVersion: apps/v1
      kind: Deployment
      name: nginx
  placement:
    clusterAffinities:
    - affinityName: primary-clusters
      clusterNames:
        - member1
    - affinityName: backup-clusters
      clusterNames:
        - member2
    clusterTolerations:
    - effect: NoExecute
      key: cluster.karmada.io/not-ready
      operator: Exists
      tolerationSeconds: 0
    - effect: NoExecute
      key: cluster.karmada.io/unreachable
      operator: Exists
      tolerationSeconds: 0
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  replicas: 2
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
        - image: nginx
          name: nginx
```

The behavior is the same as you described. I will try to find out the reason, so please give me some time.

**Comment by jwcesign** [comment]
Hi, @lorenzogio97 , If you need this emergency, u can use `purgeMode: Graciously`, it works fine.

**Comment by lorenzogio97** [comment]
@jwcesign thank you for the reply. I will try setting purgeMode: Graciously to see if it works.

**Comment by jwcesign** [comment]
@lorenzogio97, could you please provide some details on why you use `purgeMode: Immediately` and in which scenarios it is suitable?

**Comment by lorenzogio97** [comment]
@jwcesign there is no real motivation to use `Immediately `instead of `Graciously`, I was testing the feature and I chose this mode. If it works with `Graciously`, for me it's ok. I will test it today.

**Comment by RainbowMango** [comment]
@XiShanYongYe-Chang Could you please confirm if this issue has been addressed by #5788?

**Comment by XiShanYongYe-Chang** [comment]
> but when I cordon the cluster where pods are currenty scheduled and delete them, karmada continue to try to re-schedule them in the same cluster, for more than the tolerationSeconds that are expected.

This is the core issue described in the current issue. For Immediately PurgeMode, we did not add the cluster to the eviction queue during eviction. Now, we have modified its behavior so that karmada-scheduler can filter the cluster during re-scheduling.

To sum up, we've solved this problem.

**Comment by RainbowMango** [comment]
Here is PR #5881, the fix will be included in the coming v1.12.0 release which is planned at the end of this month. (2024-11-30).

/close
@lorenzogio97 Please feel free to let us know if it doesn't work.
